### PR TITLE
HIP: Add input and output wrappers for faster load and store address computation

### PR DESCRIPTION
**Pull request by Maetveis**
On HIP if all addresses from a buffer are smaller than 32 bit, then addressing calculations can be done in 32 bits saving a few instructions and registers each time a buffer is accessed. Add a new option useStrict32BitAddress for this optimization. 1 means enable, -1 disable, 0 to infer from size. Note that this can only be done if the addresses (in bytes) are < 32 bit not the indices (number of elements) therefore this is not the same as useUint64, but it has no effect when useUint64 is 1.

[See it in action on compiler explorer](https://godbolt.org/z/GjbnzT645)

I verified this locally by adding sizes to the precision test that are just below and above the 32 bit limit (e.g (2^15 - 1) x 2^15, 2^15 x 2^15).

### CODE REVIEW

File: vkFFT/vkFFT.h [hunk 1]
```diff
@@ -281,6 +281,7 @@ typedef struct {
 	hipEvent_t* stream_event;//Filled at app creation
 	uint64_t streamCounter;//Filled at app creation
 	uint64_t streamID;//Filled at app creation
+	int64_t  useStrict32BitAddress; // guarantee 32 bit addresses in bytes instead of number of elements. This results in fewer instructions generated. -1: Disable, 0: Infer based on size, 1: enable. Has no effect with useUint64.
```
Comment (DTolm) [inline vkFFT/vkFFT.h:284]: This will break all other backends as you reference it not only in VKFFT_BBACKEND=2.

File: vkFFT/vkFFT.h [hunk 2]
```diff
@@ -35409,6 +35445,7 @@ static inline VkFFTResult VkFFTPlanAxis(VkFFTApplication* app, VkFFTPlan* FFTPla
 	axis->specializationConstants.warpSize = app->configuration.warpSize;
 	axis->specializationConstants.numSharedBanks = app->configuration.numSharedBanks;
 	axis->specializationConstants.useUint64 = app->configuration.useUint64;
+	axis->specializationConstants.useStrict32BitAddress = app->configuration.useStrict32BitAddress;
```
Comment (DTolm) [inline vkFFT/vkFFT.h:35448]: This will break all other backends as you reference it not only in VKFFT_BBACKEND=2.

**Comment by DTolm** [comment]
Hello,

While I believe that this is a correct workaround around HIP compiler, this will probably not improve the performance in any meaningful way (int calculations are pretty fast). I will merge this pr, but do you have an example where this improves performance?

Also, you need to add the same wrapper in shaderGenVkFFT_R2C_decomposition function.

Also, you declare useStrict32BitAddress as HIP-only variable which will break all other backends when you reference it.

Best regards,
Dmitrii

File: vkFFT/vkFFT.h (hunk 1 above)
Comment (Maetveis) [reply vkFFT/vkFFT.h:284]: Yep, seems like I missed a few places, should be fixed now.

File: vkFFT/vkFFT.h (hunk 2 above)
Comment (Maetveis) [reply vkFFT/vkFFT.h]: Should be fixed now.

**Comment by Maetveis** [comment]
> While I believe that this is a correct workaround around HIP compiler, this will probably not improve the performance in any meaningful way (int calculations are pretty fast). I will merge this pr, but do you have an example where this improves performance?

Originally I struggled to find benchmarks where this made a difference, I was trying `Vulkan_FFT -vkkft 1`, but since then I found that for `Vulkan_FFT -vkkft 3` will show improvements particularly in small sizes:

out

This was run on MI210 with rocm 5.2.3, and I have taken a maximum of 3 runs for each size, but I'll attach the full document: 
[vkfft-3_strict_32b_indexing.ods](https://github.com/DTolm/VkFFT/files/9778229/vkfft-3_strict_32b_indexing.ods)

> Also, you need to add the same wrapper in shaderGenVkFFT_R2C_decomposition function.

Sure let me check, I'll push yet another update shortly.

> Also, you declare useStrict32BitAddress as HIP-only variable which will break all other backends when you reference it.

This should now be fixed.

**Comment by DTolm** [comment]
Small systems are extremely cache-hit dependent so you probably have noticed that the results of benchmark often differ more than the measured performance gains. Having better assembly, however, is always better.

**Pull request merged**